tracing-core/src/field/map.rs

```diff
@@ -0,0 +1,356 @@
+#![allow(missing_docs)]
+use super::{Field, FieldSet};
+use std::fmt;
+
+/// A map of `Field`s to `T`s, with fast O(1) indexing.
+///
+/// In many cases, subscriber implementations may wish to store data for each
+/// field recorded by a span or event. `FieldMap` provides specialized storage
+/// for that use-case, with `Field`s from a `FieldSet` as the key type.
+///
+/// Unlike other storage, such as `HashMap`, a `FieldMap` does not allocate, and
+/// can store all data on the stack or inline in another struct. Like `HashMap`,
+/// it has O(1) indexing, but with significantly faster constant-faster
+/// performance. However, it may _only_ be indexed by `Field`s.
+pub struct FieldMap<T> {
+    fields: FieldSet,
+    values: [Option<T>; 32],
```

Comment:
Where does the `32` come from? Are `FieldSet` limited to 32 as well? Would be good to document that fact.

Reply:
@jonhoo there's a constraint on `ValueSet` limiting it to >= 32 fields:
https://github.com/tokio-rs/tracing/blob/0a5ba16c5fea076f4902995bd59dedd99cac2755/tracing-core/src/field.rs#L493-L498

I realize just now that `FieldSet` doesn't actually have the same constraint, which _technically_ would allow constructing `FieldSet`s with more values than fit in a `ValueSet`; since the span and event macros invoke both constructors, this can't happen with the macros, but a user calling `FieldSet::new` could potentially make this mistake. We should probably update the `FieldSet` constructor to enforce that constraint as well...

Reply:
Since `FieldSet::value_set` is `doc(hidden)`, the above doc comment isn't public, but the macro docs do state this publicly:

https://github.com/tokio-rs/tracing/blob/0a5ba16c5fea076f4902995bd59dedd99cac2755/tracing/src/macros.rs#L169-L184

We should probably document the length limit in the `tracing-core` docs as well.